### PR TITLE
Implements the Ballmer Peak: Now the bartender can help R&D!

### PR DESCRIPTION
**Pull request by Iamgoofball**
See title. Here be math.

:cl: Iamgoofball
fix: Booze now properly reflects programming speed.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/life.dm [hunk 1]
```diff
@@ -368,20 +368,37 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
+		if(mind && (mind.assigned_role == "Scientist" || mind.assigned_role == "Research Director"))
+			if(SSresearch.science_tech)
+				if(drunkenness >= 12.9 && drunkenness <= 13.8)
+					var/ballmer_percent = 0
+					if(drunkenness > 13.35)
+						ballmer_percent = ((drunkenness - 13.8) / (12.9 - 13.8) * 2)
+					else if(drunkenness < 13.35)
+						ballmer_percent = ((drunkenness - 12.9) / (13.8 - 12.9) * 2)
+					else
+						ballmer_percent = 1 // right on the money
```
Comment (ninjanomnom) [inline code/modules/mob/living/carbon/human/life.dm]:
```
if(drunkenness == 13.35)
  balmer_percent = 1
else
  balmer_percent = (-abs(drunkenness - 13.35) / 0.9) + 0.5
```
https://www.desmos.com/calculator/hyftwkt2kw

Comment (Iamgoofball) [reply code/modules/mob/living/carbon/human/life.dm]: oh shit thanks, was struggling with that

Comment (ninjanomnom) [reply code/modules/mob/living/carbon/human/life.dm]: You may also want to round drunkenness

**Comment by ninjanomnom** [comment]
Do we have a good way of learning the durnkenness var? If not you should probably add that or make some way of figuring out when you're at/near the peak

**Comment by Iamgoofball** [comment]
thats cheating and part of the ballmer peak experience

**Comment by ninjanomnom** [comment]
An status effect that shows when you're gaining points would be good enough probably, otherwise this feature will just be unknown.

**Review by Tlaltecuhtli — APPROVED** [review]

**Comment by 81Denton** [comment]
I'm looking forward to seeing RnD full of shambling, alcoholic wrecks because the RD wants those points.

**Comment by Iamgoofball** [comment]
Sounds like a proper r&d environment

**Comment by MoreRobustThanYou** [comment]
@Cyberboss you better merge this. or else.

**Comment by praisenarsie** [comment]
who said cyber has to merge it

**Comment by QualityVan** [comment]
PR doesn't make you shout DEVELOPERS repeatedly, cannot approve.

**Comment by 81Denton** [comment]
>PR doesn't make you shout DEVELOPERS repeatedly, cannot approve.
>i might add a notification if you're getting peak bonuses

If you don't start shouting DEVELOPERS when hitting the Ballmer peak I'll be disappointed.

**Comment by MoreRobustThanYou** [comment]
DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS, DEVELOPERS.... x2000

File: code/modules/mob/living/carbon/human/life.dm [hunk 2]
```diff
@@ -368,20 +368,35 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
+						ballmer_percent = 1
+					else
+						ballmer_percent = (-abs(drunkenness - 13.35) / 0.9) + 0.5
+
+					if(ballmer_percent > 1)
+						ballmer_percent = 1
+					SSresearch.science_tech.research_points += (BALLMER_POINTS * ballmer_percent)
+				if(drunkenness > 26) // by this point you're into windows ME territory
+					if(prob(5))
+						SSresearch.science_tech.research_points -= 1
```
Comment (MoreRobustThanYou) [inline code/modules/mob/living/carbon/human/life.dm]: 👎 from me until this makes you shout ";DEVELOPERS DEVELOPERS DEVELOPERS, DEVELOPERS..."

**Comment by outbools** [comment]
PR that saved techwebs

File: code/modules/mob/living/carbon/human/life.dm [hunk 3]
```diff
@@ -368,20 +368,35 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
+		if(mind && (mind.assigned_role == "Scientist" || mind.assigned_role == "Research Director"))
+			if(SSresearch.science_tech)
+				if(drunkenness >= 12.9 && drunkenness <= 13.8)
+					var/ballmer_percent = 0
+					if(drunkenness == 13.35) // why run math if I dont have to
+						ballmer_percent = 1
+					else
+						ballmer_percent = (-abs(drunkenness - 13.35) / 0.9) + 0.5
+
+					if(ballmer_percent > 1)
```
Comment (lordpidey) [inline code/modules/mob/living/carbon/human/life.dm]: Why?  According to the math, the highest balmer percent possible is 0.5, unless you are exactly at 13.35, there's no way for it to be above 1.

File: code/modules/mob/living/carbon/human/life.dm [hunk 4]
```diff
@@ -368,20 +368,35 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
 			jitteriness = max(jitteriness - 3, 0)
 
 		if(drunkenness >= 11 && slurring < 5)
 			slurring += 1.2
-
+		if(mind && (mind.assigned_role == "Scientist" || mind.assigned_role == "Research Director"))
+			if(SSresearch.science_tech)
+				if(drunkenness >= 12.9 && drunkenness <= 13.8)
+					var/ballmer_percent = 0
+					if(drunkenness == 13.35) // why run math if I dont have to
```
Comment (lordpidey) [inline code/modules/mob/living/carbon/human/life.dm]: A == comparison on a floating point number?  There's a significant chance that that will LITERALLY never trigger.  Make it a range.

**Review by lordpidey — CHANGES_REQUESTED** [review]
The particulars of the math have some issues.

File: code/modules/mob/living/carbon/human/life.dm (hunk 3 above)
Comment (Iamgoofball) [reply code/modules/mob/living/carbon/human/life.dm]: Figured I'd keep it as insurance

File: code/modules/mob/living/carbon/human/life.dm (hunk 4 above)
Comment (Iamgoofball) [reply code/modules/mob/living/carbon/human/life.dm]: That's intentional.

Comment (ninjanomnom) [reply code/modules/mob/living/carbon/human/life.dm]: With this few significant digits it'll trigger just fine even as a floating point number. As I said in my above comment though it should definitely round drunkenness.

File: code/modules/mob/living/carbon/human/life.dm (hunk 3 above)
Comment (ninjanomnom) [reply code/modules/mob/living/carbon/human/life.dm]: There's really no need

File: code/modules/mob/living/carbon/human/life.dm (hunk 2 above)
Comment (lordpidey) [inline code/modules/mob/living/carbon/human/life.dm]: This adds a hilarious way for neer-do-wells to slow down research using only a syringe gun.

Comment (Iamgoofball) [reply code/modules/mob/living/carbon/human/life.dm]: sounds like emergent gameplay

**Comment by MoreRobustThanYou** [comment]
relevant: https://www.youtube.com/watch?v=KMU0tzLwhbE

**Comment by deathride58** [comment]
Holy shit you actually implemented it you absolute madman

**Comment by Iamgoofball** [comment]
can someone restart travis, the final one apparently had a travis die

**Comment by Iamgoofball** [comment]
this actually does compile, travis just died if you check the logs

**Review by ninjanomnom — DISMISSED** [review]
There needs to be some indication to the player that they're getting points

**Comment by JStheguy** [comment]
Hitting the Balmer Peak makes Scientists randomly yell "Hey, guys! What if we [do something smart]?"
Hitting Windows ME territory makes Scientists randomly say "Hey, guys. Hey, w-what if we, uh, what if we [do literally the worst thing we could possibly do to fuck everything up at the moment]?" and then start vomiting on the R&D console.

**Comment by arsserpentarium** [comment]
Oh ,now I know, how supermatter sword was invented!

**Comment by Cyberboss** [comment]
Can't tell if mso is adding this label or this is some fuckup in the bot...

**Comment by Spaceinaba** [comment]
goof first pr

**Comment by ExcessiveUseOfCobblestone** [comment]
> There needs to be some indication to the player that they're getting points

Does that not encourage people minmaxing? How do you plan on curbing that?

**Comment by Iamgoofball** [comment]
@ExcessiveUseOfCobblestone idk what you're talking about, they're talking about the code they researched in game and the deployment R&D server that passes info to the departmental lathes

what do you think R&D does, make everything in Scratch?

File: code/modules/mob/living/carbon/human/life.dm [hunk 5]
```diff
@@ -368,20 +368,45 @@ All effects don't start immediately, but rather get worse over time; the rate is
...
+										"Hear me out here. What if, and this is just a theory, we made R&D controllable from our PDAs?",
+										"I'm thinking we should roll out a git repository for our research under the AGPLv3 license so that we can share it among the other stations freely.",
+										"I dunno about you guys, but IDs and PDAs being separate is clunky as fuck. Maybe we should merge them into a chip in our arms? That way they can't be stolen easily.",
+										"Why the fuck aren't we just making every pair of shoes into galoshes? We have the technology."))
+GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put a webserver that's automatically turned on with default admin passwords into every PDA?",
+												"So like, you know how we separate our codebase from the master copy that runs on our consumer boxes? What if we merged the two and undid the separation between codebase and server?",
+												"Dude, radical idea: H.O.N.K mechs but with no bananium required.",
+												"Best idea ever: Disposal pipes instead of hallways."))
```
Comment (deathride58) [inline code/modules/mob/living/carbon/human/life.dm]: Move ballmer messages to config.

**Review by deathride58 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/life.dm (hunk 5 above)
Comment (MoreRobustThanYou) [reply code/modules/mob/living/carbon/human/life.dm]: move it to the strings/ folder

Comment (ninjanomnom) [reply code/modules/mob/living/carbon/human/life.dm]: we already talked about this with him in coderbus, he'll be switching it later

**Comment by Rinoka** [comment]
goof this is probably your first good pr

continue

File: code/modules/mob/living/carbon/human/life.dm (hunk 5 above)
Comment (PKPenguin321) [reply code/modules/mob/living/carbon/human/life.dm]: These are great regardless

**Comment by PKPenguin321** [comment]
Lovely

**Comment by KieranCat** [comment]
Amazing, a goof meme PR I can get behind! Actually makes the departments have interaction between them without forcing it, but still keeping options open.

**Pull request merged**